src/parser/druid/balance/modules/talents/azeritetraits/ArcanicPulsar.tsx

```diff
@@ -0,0 +1,92 @@
+import React from 'react';
+import SPELLS from 'common/SPELLS/index';
+
+import Analyzer, { SELECTED_PLAYER } from 'parser/core/Analyzer';
+
+import Statistic from 'interface/statistics/Statistic';
+import Events, { CastEvent, DamageEvent } from 'parser/core/Events';
+import STATISTIC_ORDER from 'interface/others/STATISTIC_ORDER';
+import BoringSpellValueText
+  from 'interface/statistics/components/BoringSpellValueText';
+import ItemDamageDone from 'interface/ItemDamageDone';
+import { calculateAzeriteEffects } from 'common/stats';
+
+class ArcanicPulsar extends Analyzer {
+  /**
+   * Starsurge's damage is increased by 2181. Every 9 Starsurges, gain
+   * Celestial Alignment for 6 sec.
+   *
+   * Example Log:
+   * https://www.warcraftlogs.com/reports/ZfcqAWCn7gHtp49G#fight=17&type=auras&ability=287790
+   *
+   */
+
+  protected damageGained = 0;
+  protected bonusDamage = 0;
+  protected firstStarsurgeCast = true;
+  protected celestialAlignmentCount = 0;
+
+  constructor(options: any) {
+    super(options);
+    this.active
+      = this.selectedCombatant.hasTrait(SPELLS.ARCANIC_PULSAR_TRAIT.id);
+
+    if (!this.active) {
+      return;
+    }
+
+    this.bonusDamage
+      = this.selectedCombatant.traitsBySpellId[SPELLS.ARCANIC_PULSAR_TRAIT.id]
+      .reduce((total, rank) => {
+        const [damage] = calculateAzeriteEffects(
+          SPELLS.ARCANIC_PULSAR_TRAIT.id,
+          rank,
+        );
+        return total + damage;
+      }, 0);
+
+    this.addEventListener(
+      Events.cast.by(SELECTED_PLAYER)
+        .spell(SPELLS.STARSURGE_MOONKIN),
+      this.onStarsurgeCast,
+    );
+
+    this.addEventListener(
+      Events.damage.by(SELECTED_PLAYER)
+        .spell(SPELLS.STARSURGE_MOONKIN),
+      this.onStarsurgeDamage,
+    );
+  }
+
+  onStarsurgeCast(event: CastEvent) {
+    if (this.firstStarsurgeCast) {
+      this.firstStarsurgeCast = false;
+      return;
+    }
+
+    if (!this.selectedCombatant.hasBuff(SPELLS.ARCANIC_PULSAR_BUFF.id)) {
+      this.celestialAlignmentCount++;
+    }
```

Comment:
Isn't this wrong? That's not how many times Celestial Alignment was triggered? 
Shouldn't this be tracked through a apply_buff event?

Reply:
Considering the Arcanic Pulsar buff is applied starting at one stack, and dissapears at 9 stacks. It would indeed be better to use a applybuff or applybuffstack. However the issue is that the buff will never reach 9 stacks, instead it will just disappear when you cast Starsurge at 8 stacks. And Celestial Alignment will be activated.

The best way imo would be to check if the buff has 8 stacks when Starsurge is cast however, [this.selectedCombatan.getBuff(...).stacks](url) will return an error, even though the console.log output shows its part of the object.

Reply:
Maybe i should use: floor(STARSURGE_CAST_COUNT / 9) ?

Reply:
Does the applybuffstack not work for this?

Reply:
getBuff(SPELLS).stacks will error if they dont have the buff on them. 

So you can do
buff = this.selectedCombatant.getBuff(SPELLS)
if (buff && buff.stacks > x)

But what exactly are you trying to get here? The total number of Celestial Alignment Buffs? (Casts and Procs), or arcanic pulsar procs?? Your variable is saying one thing and your if is saying another

Reply:
After discussing with abelito on how to approach this way in the best manner, i've determined that storing the stacks is the best option in order to avoid collision with regular CA casts and VOP procs.

Reply:
To explain how this works: there are two primary times you *wont* have the Arcanic Pulsar buff when you cast starsurge (SS): the first SS of the fight, and the first SS you cast right after getting the celestial alignment (CA) proc from the trait. With the firststarsurgecast boolean, the first case is eliminated, and you are left with only the casts after you get a proc. Thus, you can increment your CA proc counter. There are some other fringe cases that aren't accounted for here that would lead to some inaccuracies.. for example, getting battle res'd (in which case this would count an *extra* proc), or the fight/phase ending before you cast SS again after getting the proc (in which case this would count one too few procs).

All in all it's a pretty backwards way logically to solve the problem, but I appreciate the ingenuity of the workaround :) That said, Sharrq is probably correct in saying the error you were getting originally was just that you weren't checking for the buff actually existing before trying to get the stack count. If that is the case, going back to the original idea will be far far cleaner and clearer:

> The best way imo would be to check if the buff has 8 stacks when Starsurge is cast however, this.selectedCombatan.getBuff(...).stacks has thus far always provided me with an error

Reply:
> To explain how this works: there are two primary times you _wont_ have the Arcanic Pulsar buff when you cast starsurge (SS): the first SS of the fight, and the first SS you cast right after getting the celestial alignment (CA) proc from the trait. With the firststarsurgecast boolean, the first case is eliminated, and you are left with only the casts after you get a proc. Thus, you can increment your CA proc counter. There are some other fringe cases that aren't accounted for here that would lead to some inaccuracies.. for example, getting battle res'd (in which case this would count an _extra_ proc), or the fight/phase ending before you cast SS again after getting the proc (in which case this would count one too few procs).

I think i've already worked around all of the possible issues that could occour with having an in accurate reading of the AP stacks and CA procs. Including the inaccuracy of being CRed

> All in all it's a pretty backwards way logically to solve the problem, but I appreciate the ingenuity of the workaround :) That said, Sharrq is probably correct in saying the error you were getting originally was just that you weren't checking for the buff actually existing before trying to get the stack count. If that is the case, going back to the original idea will be far far cleaner and clearer:

Interestingly enough i also tought that the error was happening because the buff = undefined at some times. And had already checked if the buff existed in the same way Sharqq proposed. However the error seemed to presist; http://prntscr.com/segzc5

Reply:
Do you have a log for this?

Reply:
Glad this got resolved! We figured out it was just a Typescript error...

Reply:
Thanks again for your time! Removed some unneccesary code <3